### PR TITLE
Add feature - find r/module-role

### PR DESCRIPTION
**Pull request by huangjiaxi1111**
Find command now support finding contacts based on module-role with prefix r/, e.g. find r/cs1101s-ta.
If role is not specified, student is assumed.
When multiple inputs are given, e.g. find r/cs1101s-ta r/cs1231s n/Ben, person matching both the name keywords and module-role keywords will be shown. This command finds person(s) named Ben AND taking cs1101s OR cs1231s.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 1]
```diff
@@ -1,50 +1,59 @@
...
+        this.predicates = predicates;
     }
 
     @Override
     public CommandResult execute(Model model) {
         requireNonNull(model);
-        model.updateFilteredPersonList(predicate);
+        predicates.forEach(model::updateFilteredPersonList);
+        String formattedKeywords = getAllKeywordsFromPredicates(predicates);
+
```
Comment (Elinengu) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:52]: I think the reason why the find command only works for 1 predicate (although I'm not sure why it only works for module but not name) is the way that predicates being passed into the ``model::updateFilteredPersonList``. Maybe you can try to combine the predicates before passing it into the method?

File: src/main/java/seedu/address/model/person/ModuleRoleMap.java [hunk 2]
```diff
@@ -177,4 +178,21 @@ public List<ModuleRolePair> getData() {
...
+     * @return a list of module-role pairs.
+     */
+    public List<String> toModuleRolePairs() {
+        List<String> moduleRolePairs = new ArrayList<>();
+
+        for (Map.Entry<ModuleCode, RoleType> entry : roles.entrySet()) {
+            ModuleCode moduleCode = entry.getKey();
+            RoleType roleType = entry.getValue();
+            moduleRolePairs.add(moduleCode + "-" + roleType);
+        }
```
Comment (Elinengu) [inline src/main/java/seedu/address/model/person/ModuleRoleMap.java:194]: Perhaps we can make use of the toString method in ModuleRolePair? maybe we can convert it to List<ModuleRolePair> using getData() then invoke the toString method and make it a list to make it looks like being written by a same person (?)

**Review by Elinengu — COMMENTED** [review]
Thank you for your hard work in writing succinct code and adding comprehensive test cases to our codebase. I have made some suggestions in the code review, you could take a look on it.

File: src/main/java/seedu/address/logic/commands/FindCommand.java (hunk 1 above)
Comment (Elinengu) [reply src/main/java/seedu/address/logic/commands/FindCommand.java:52]: image
Maybe Predicate::or could be useful here

Comment (huangjiaxi1111) [reply src/main/java/seedu/address/logic/commands/FindCommand.java:52]: Thank you for pointing out! I think this is the correct way of doing it!

File: src/test/java/seedu/address/logic/commands/FindCommandTest.java [hunk 3]
```diff
@@ -5,87 +5,172 @@
...
+        List<Predicate<Person>> secondModuleRolePredicates = new ArrayList<>();
+        secondModuleRolePredicates.add(secondModuleRolePredicate);
+
+        List<Predicate<Person>> firstNameAndModuleRolePredicates = new ArrayList<>();
+        firstNameAndModuleRolePredicates.add(firstNamePredicate);
+        firstNameAndModuleRolePredicates.add(firstModuleRolePredicate);
+
+        List<Predicate<Person>> secondNameAndModuleRolePredicates = new ArrayList<>();
+        secondNameAndModuleRolePredicates.add(secondNamePredicate);
+        secondNameAndModuleRolePredicates.add(secondModuleRolePredicate);
```
Comment (weiseng18) [inline src/test/java/seedu/address/logic/commands/FindCommandTest.java:69]: can consider using `Arrays.asList()` to define the contents of the `List` inline

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 4]
```diff
@@ -1,50 +1,60 @@
...
+            + PREFIX_MODULE + "CS2030S";
 
-    private final NameContainsKeywordsPredicate predicate;
+    private final List<Predicate<Person>> predicates;
 
-    public FindCommand(NameContainsKeywordsPredicate predicate) {
-        this.predicate = predicate;
+    public FindCommand(List<Predicate<Person>> predicates) {
+        this.predicates = predicates;
     }
```
Comment (weiseng18) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:43]: from our usage, we expect either 1 or 2 Predicates in the list, so maybe can add an assert statement for defensive programming?

Comment (weiseng18) [reply src/main/java/seedu/address/logic/commands/FindCommand.java:43]: (an alternative is to use `Optional`, but I think that would take a bit too much time for refactoring)

**Review by weiseng18 — COMMENTED** [review]
seems good overall! i've added some code nits as comments

one more thing, i think it might be good to update the command result message to be a bit more meaningful

taking the `findNameAndModuleExample.png` as an example,

image

i think it would be good to convey that it's (john OR ben) AND (MA1522 or CS1101S)

**Review by weiseng18 — APPROVED** [review]
nice!

**Pull request merged**